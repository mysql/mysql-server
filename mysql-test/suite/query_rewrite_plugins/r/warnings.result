#
# Test of rules where the parser gives warnings about the originial
# query, pattern or replacement.
#
call mtr.add_suppression("Plugin rewriter reported: 'Some rules failed to load.'");
# Query rewrite plugin was installed.
INSERT INTO query_rewrite.rewrite_rules ( pattern, replacement )
VALUES ( 'LOAD DATA INFILE "tab1.data" INTO TABLE test.t1',
'LOAD DATA INFILE "tab2.data" INTO TABLE test.t1' ),
( 'DO CONCAT("a", "b")', 'DO CONCAT("c", "d")' );
CALL query_rewrite.flush_rewrite_rules();
ERROR 45000: Loading of some rule(s) failed.
SELECT * FROM query_rewrite.rewrite_rules;
id	pattern	pattern_database	replacement	enabled	message
1	LOAD DATA INFILE "tab1.data" INTO TABLE test.t1	NULL	LOAD DATA INFILE "tab2.data" INTO TABLE test.t1	YES	Statement type of pattern not supported.
2	DO CONCAT("a", "b")	NULL	DO CONCAT("c", "d")	YES	Statement type of pattern not supported.
INSERT INTO query_rewrite.rewrite_rules ( pattern, replacement )
VALUES ( 'SHOW WARNINGS', 'SHOW /* I rewrote it */ WARNINGS' );
CALL query_rewrite.flush_rewrite_rules();
ERROR 45000: Loading of some rule(s) failed.
# We want the note about the query being rewritten here ...
SHOW WARNINGS;
Level	Code	Message
Error	1644	Loading of some rule(s) failed.
SHOW ERRORS;
Level	Code	Message
Error	1644	Loading of some rule(s) failed.
GET DIAGNOSTICS CONDITION 1 @p1 = RETURNED_SQLSTATE, @p2 = MESSAGE_TEXT;
SELECT @p1;
@p1
45000
SELECT @p2;
@p2
Loading of some rule(s) failed.
# ... but not here, since we failed parsing after seeing SHOW WARNINGS.
SHOW WARNINGS and give me a parse error, please;
ERROR 42000: You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near 'and give me a parse error, please' at line 1
DELETE FROM query_rewrite.rewrite_rules;
# Test of warnings when we rewrite the diagnostics statement.  echo #
# We disable mtr's SHOW WARNINGS here, as it would just be confusing
# when it gets rewritten without us seeing it.
INSERT INTO query_rewrite.rewrite_rules ( pattern, replacement )
VALUES ( 'SHOW WARNINGS', 'SELECT "used to be SHOW WARNINGS"' ),
( 'SELECT 1', 'SHOW WARNINGS' );
CALL query_rewrite.flush_rewrite_rules();
ERROR 45000: Loading of some rule(s) failed.
SELECT 1;
Level	Code	Message
Error	1644	Loading of some rule(s) failed.
Note	1105	Query 'SELECT 1' rewritten to 'SHOW WARNINGS' by a query rewrite plugin
SHOW WARNINGS;
Level	Code	Message
Error	1644	Loading of some rule(s) failed.
Note	1105	Query 'SELECT 1' rewritten to 'SHOW WARNINGS' by a query rewrite plugin
Warnings:
Warning	1620	Plugin is busy and will be uninstalled on shutdown
# Query rewrite plugin was queued for uninstalling.
