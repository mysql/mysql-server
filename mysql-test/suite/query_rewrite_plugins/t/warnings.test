--echo #
--echo # Test of rules where the parser gives warnings about the originial
--echo # query, pattern or replacement.
--echo #

call mtr.add_suppression("Plugin rewriter reported: 'Some rules failed to load.'");

--source suite/query_rewrite_plugins/include/have_plugin_rewriter.inc
--source suite/query_rewrite_plugins/include/install_rewriter.inc

# Still many unsupported statements
INSERT INTO query_rewrite.rewrite_rules ( pattern, replacement )
VALUES ( 'LOAD DATA INFILE "tab1.data" INTO TABLE test.t1',
         'LOAD DATA INFILE "tab2.data" INTO TABLE test.t1' ),
       ( 'DO CONCAT("a", "b")', 'DO CONCAT("c", "d")' );

--error ER_SIGNAL_EXCEPTION
CALL query_rewrite.flush_rewrite_rules();
SELECT * FROM query_rewrite.rewrite_rules;

INSERT INTO query_rewrite.rewrite_rules ( pattern, replacement )
VALUES ( 'SHOW WARNINGS', 'SHOW /* I rewrote it */ WARNINGS' );

--error ER_SIGNAL_EXCEPTION
CALL query_rewrite.flush_rewrite_rules();

--echo # We want the note about the query being rewritten here ...
SHOW WARNINGS;
SHOW ERRORS;
GET DIAGNOSTICS CONDITION 1 @p1 = RETURNED_SQLSTATE, @p2 = MESSAGE_TEXT;
SELECT @p1;
SELECT @p2;

--echo # ... but not here, since we failed parsing after seeing SHOW WARNINGS.
--error ER_PARSE_ERROR
SHOW WARNINGS and give me a parse error, please;

DELETE FROM query_rewrite.rewrite_rules;

--echo # Test of warnings when we rewrite the diagnostics statement.  echo #
--echo # We disable mtr's SHOW WARNINGS here, as it would just be confusing
--echo # when it gets rewritten without us seeing it.
--disable_warnings
INSERT INTO query_rewrite.rewrite_rules ( pattern, replacement )
VALUES ( 'SHOW WARNINGS', 'SELECT "used to be SHOW WARNINGS"' ),
       ( 'SELECT 1', 'SHOW WARNINGS' );

--error ER_SIGNAL_EXCEPTION
CALL query_rewrite.flush_rewrite_rules();

SELECT 1;

SHOW WARNINGS;

--enable_warnings
--source suite/query_rewrite_plugins/include/uninstall_rewriter.inc
