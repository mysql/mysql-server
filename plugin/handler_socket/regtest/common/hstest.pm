
# vim:sw=2:ai

package hstest;

use DBI;
use Net::HandlerSocket;

our %conf = ();

sub get_conf_env {
  my ($key, $defval) = @_;
  return $ENV{$key} || $defval;
}

sub init_conf {
  $conf{host} = get_conf_env("MYHOST", "localhost");
  $conf{myport} = get_conf_env("MYPORT", 3306);
  $conf{dbname} = get_conf_env("MYDBNAME", "hstestdb");
  $conf{ssps} = get_conf_env("MYSSPS");
  $conf{user} = get_conf_env("MYSQLUSER", "root");
  $conf{pass} = get_conf_env("MYSQLPASS", "");
  $conf{hsport} = get_conf_env("HSPORT", 9998);
  $conf{hspass} = get_conf_env("HSPASS", undef);
}

sub get_dbi_connection {
  my ($dbname, $host, $myport, $ssps, $user, $pass)
    = ($conf{dbname}, $conf{host}, $conf{myport}, $conf{ssps},
      $conf{user}, $conf{pass});
  my $mycnf = "binary_my.cnf";
  my $dsn = "DBI:mysql:database=;host=$host;port=$myport"
    . ";mysql_server_prepare=$ssps"
    . ";mysql_read_default_group=perl"
    . ";mysql_read_default_file=../common/$mycnf";
  my $dbh = DBI->connect($dsn, $user, $pass, { RaiseError => 1 });
  return $dbh;
}

sub init_testdb {
  my $charset = $_[0] || "binary";
  my $dbh = get_dbi_connection();
  my $dbname = $conf{dbname};
  $dbh->do("drop database if exists $dbname");
  $dbh->do("create database $dbname default character set $charset");
  $dbh->do("use $dbname");
  return $dbh;
}

sub get_hs_connection {
  my ($host, $port) = @_;
  $host ||= $conf{host};
  $port ||= $conf{hsport};
  my $hsargs = { 'host' => $host, 'port' => $port };
  my $conn = new Net::HandlerSocket($hsargs);
  if (defined($conn) && defined($conf{hspass})) {
    $conn->auth($conf{hspass});
  }
  return $conn;
}


init_conf();

1;

