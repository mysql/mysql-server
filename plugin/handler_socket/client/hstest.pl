#!/usr/bin/perl

# vim:sw=8:ai:ts=8

use strict;
use warnings;

use DBI;
use Net::HandlerSocket;

my %conf = ();
for my $i (@ARGV) {
	my ($k, $v) = split(/=/, $i);
	$conf{$k} = $v;
}

my $verbose = get_conf("verbose", 0);
my $actions_str = get_conf("actions", "hsread");
my $tablesize = get_conf("tablesize", 10000);
my $db = get_conf("db", "hstest");
my $table = get_conf("table", "hstest_table1");
my $engine = get_conf("engine", "innodb");
my $host = get_conf("host", "localhost");
my $mysqlport = get_conf("mysqlport", 3306);
my $mysqluser = get_conf("mysqluser", "root");
my $mysqlpass = get_conf("mysqlpass", "");
my $hsport = get_conf("hsport", 9999);
my $loop = get_conf("loop", 10000);
my $op = get_conf("op", "=");
my $ssps = get_conf("ssps", 0);
my $num_moreflds = get_conf("moreflds", 0);
my $moreflds_prefix = get_conf("moreflds_prefix", "column0123456789_");
my $keytype = get_conf("keytype", "varchar(32)");
my $file = get_conf("file", undef);

my $dsn = "DBI:mysql:database=;host=$host;port=$mysqlport"
	. ";mysql_server_prepare=$ssps";
my $dbh = DBI->connect($dsn, $mysqluser, $mysqlpass, { RaiseError => 1 });
my $hsargs = { 'host' => $host, 'port' => $hsport };
my $cli = new Net::HandlerSocket($hsargs);

my @actions = split(/,/, $actions_str);
for my $action (@actions) {
	if ($action eq "table") {
		print("TABLE $db.$table\n");
		$dbh->do("drop database if exists $db");
		$dbh->do("create database $db");
		$dbh->do("use $db");
		my $moreflds = get_createtbl_moreflds_str();
		$dbh->do(
			"create table $table (" .
			"k $keytype primary key" .
			",v varchar(32) not null" .
			$moreflds .
			") character set utf8 collate utf8_bin " .
			"engine = $engine");
	} elsif ($action eq "insert") {
		print("INSERT $db.$table tablesize=$tablesize\n");
		$dbh->do("use $db");
		my $moreflds = get_insert_moreflds_str();
		for (my $i = 0; $i < $tablesize; $i += 100) {
			my $qstr = "insert into $db.$table values";
			for (my $j = 0; $j < 100; ++$j) {
				if ($j != 0) {
					$qstr .= ",";
				}
				my $k = "" . ($i + $j);
				my $v = "v" . int(rand(1000)) . ($i + $j);
				$qstr .= "('$k', '$v'";
				for (my $j = 0; $j < $num_moreflds; ++$j) {
					$qstr .= ",'$j'";
				}
				$qstr .= ")";
			}
			$dbh->do($qstr);
			print "$i/$tablesize\n" if $i % 1000 == 0;
		}
	} elsif ($action eq "read") {
		print("READ $db.$table op=$op loop=$loop\n");
		$dbh->do("use $db");
		my $moreflds = get_select_moreflds_str();
		my $sth = $dbh->prepare(
			"select k,v$moreflds from $db.$table where k = ?");
		for (my $i = 0; $i < $loop; ++$i) {
			my $k = "" . int(rand($tablesize));
			# print "k=$k\n";
			$sth->execute($k);
			if ($verbose >= 10) {
				print "RET:";
				while (my $ref = $sth->fetchrow_arrayref()) {
					my $rk = $ref->[0];
					my $rv = $ref->[1];
					print " $rk $rv";
				}
				print "\n";
			}
			print "$i/$loop\n" if $i % 1000 == 0;
		}
	} elsif ($action eq "hsinsert") {
		print("HSINSERT $db.$table tablesize=$tablesize\n");
		$cli->open_index(1, $db, $table, '', 'k,v');
		for (my $i = 0; $i < $tablesize; ++$i) {
			my $k = "" . $i;
			my $v = "v" . int(rand(1000)) . $i;
			my $r = $cli->execute_insert(1, [ $k, $v ]);
			if ($r->[0] != 0) {
				die;
			}
			print "$i/$tablesize\n" if $i % 1000 == 0;
		}
	} elsif ($action eq "hsread") {
		print("HSREAD $db.$table op=$op loop=$loop\n");
		my $moreflds = get_select_moreflds_str();
		$cli->open_index(1, $db, $table, '', "k,v$moreflds");
		for (my $i = 0; $i < $loop; ++$i) {
			my $k = "" . int(rand($tablesize));
			# print "k=$k\n";
			my $r = $cli->execute_find(1, $op, [ $k ], 1, 0);
			if ($verbose >= 10) {
				my $len = scalar(@{$r});
				print "LEN=$len";
				for my $e (@{$r}) {
					print " [$e]";
				}
				print "\n";
			}
			print "$i/$loop\n" if $i % 1000 == 0;
		}
	} elsif ($action eq "hsupdate") {
		my $vbase = "v" . int(rand(1000));
		print("HSUPDATE $db.$table op=$op loop=$loop vbase=$vbase\n");
		$cli->open_index(1, $db, $table, '', 'v');
		for (my $i = 0; $i < $loop; ++$i) {
			my $k = "" . int(rand($tablesize));
			my $v = $vbase . $i;
			print "k=$k v=$v\n";
			my $r = $cli->execute_update(1, $op, [ $k ], 1, 0,
				[ $v ]);
			if ($verbose >= 10) {
				print "UP k=$k v=$v\n";
			}
			print "$i/$loop\n" if $i % 1000 == 0;
		}
	} elsif ($action eq "hsdelete") {
		print("HSDELETE $db.$table op=$op loop=$loop\n");
		$cli->open_index(1, $db, $table, '', '');
		for (my $i = 0; $i < $loop; ++$i) {
			my $k = "" . int(rand($tablesize));
			print "k=$k\n";
			my $r = $cli->execute_delete(1, $op, [ $k ], 1, 0);
			if ($verbose >= 10) {
				print "DEL k=$k\n";
			}
			print "$i/$loop\n" if $i % 1000 == 0;
		}
	} elsif ($action eq "verify") {
		verify_do();
	}
}

sub verify_do {
	my ($fail_cnt, $ok_cnt) = (0, 0);
	my $sth = $dbh->prepare("select v from $db.$table where k = ?");
	use FileHandle;
	my $fh = new FileHandle($file, "r");
	while (my $line = <$fh>) {
		chomp($line);
		my @vec = split(/\t/, $line);
		my $k = $vec[3];
		my $v = $vec[7];
		next if (!defined($k) || !defined($v));
		# print "$k $v\n";
		$sth->execute($k);
		my $aref = $sth->fetchrow_arrayref();
		if (!defined($aref)) {
			print "FAILED: $k notfound\n";
			++$fail_cnt;
		} else {
			my $gv = $aref->[0];
			if ($gv ne $v) {
				print "FAILED: $k got=$gv expected=$v\n";
				++$fail_cnt;
			} else {
				print "OK: $k $v $gv\n" if $verbose >= 10;
				++$ok_cnt;
			}
		}
	}
	print "OK=$ok_cnt FAIL=$fail_cnt\n";
}

sub get_conf {
	my ($key, $def) = @_;
	my $val = $conf{$key};
	if ($val) {
		print "$key=$val\n";
	} else {
		$val = $def;
		$def ||= '';
		print "$key=$def(default)\n";
	}
	return $val;
}

sub get_createtbl_moreflds_str {
	my $s = "";
	for (my $j = 0; $j < $num_moreflds; ++$j) {
		$s .= ",$moreflds_prefix$j varchar(30)";
	}
	return $s;
}

sub get_select_moreflds_str {
	my $s = "";
	for (my $i = 0; $i < $num_moreflds; ++$i) {
		$s .= ",$moreflds_prefix$i";
	}
	return $s;
}

sub get_insert_moreflds_str {
	my $s = "";
	for (my $i = 0; $i < $num_moreflds; ++$i) {
		$s .= ",?";
	}
	return $s;
}

