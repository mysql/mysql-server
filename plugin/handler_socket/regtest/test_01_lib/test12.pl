#!/usr/bin/perl

# vim:sw=2:ai

# test for filters

BEGIN {
	push @INC, "../common/";
};

use strict;
use warnings;
use hstest;

my $dbh = hstest::init_testdb();
my $table = 'hstesttbl';
my $tablesize = 10;
$dbh->do(
  "create table $table " .
  "(k1 varchar(30) not null, k2 varchar(30) not null, " .
  "v1 int not null, v2 int not null, " .
  "primary key (k1, k2) ) engine = innodb");
srand(999);

my $sth = $dbh->prepare("insert into $table values (?,?,?,?)");
for (my $i = 0; $i < $tablesize; ++$i) {
  for (my $j = 0; $j < $tablesize; ++$j) {
    my $k1 = "k1_" . $i;
    my $k2 = "k2_" . $j;
    my $v1 = $i + $j;
    my $v2 = $i * $j;
    $sth->execute($k1, $k2, $v1, $v2);
  }
}

my $hs = hstest::get_hs_connection(undef, 9999);
my $dbname = $hstest::conf{dbname};
$hs->open_index(1, $dbname, $table, '', 'k1,k2,v1,v2', 'k2');
$hs->open_index(2, $dbname, $table, '', 'k1,k2,v1,v2', 'k1,k2,v1,v2');
$hs->open_index(3, $dbname, $table, '', 'v1', 'k1,v2');

exec_multi(
  4, "VAL",
  [ 1, '>=', [ '', '' ], 1000, 0 ],
);
# all

# select k1, k2, v1, v2 ... where (k1, k2) >= ('', '') and k2 = 'k2_5'
exec_single(
  4, "FILTER",
  [ 1, '>=', [ '', '' ], 1000, 0, undef, undef, [ [ 'F', '=', 0, 'k2_5' ] ] ]
);

# same as above
exec_multi(
  4, "FILTER",
  [ 1, '>=', [ '', '' ], 1000, 0, undef, undef, [ [ 'F', '=', 0, 'k2_5' ] ] ],
);

# select k1, k2, v1, v2 ... where (k1, k2) >= ('', '') and v1 = 3
exec_multi(
  4, "FILTER",
  [ 2, '>=', [ '', '' ], 1000, 0, undef, undef, [ [ 'F', '=', 2, 3 ] ] ],
);

# select k1, k2, v1, v2 ... where (k1, k2) >= ('k1_1', '') and k1 <= 'k1_2'
exec_multi(
  4, "FILTER",
  [ 2, '>=', [ 'k1_1', '' ], 1000, 0, undef, undef,
    [ [ 'W', '<=', 0, 'k1_2' ] ] ],
);

# select k1, k2, v1, v2 ... where (k1, k2) >= ('k1_1', '') and k1 <= 'k1_2'
#   and v2 >= 10
exec_multi(
  4, "FILTER",
  [ 2, '>=', [ 'k1_1', '' ], 1000, 0, undef, undef,
    [ [ 'W', '<=', 0, 'k1_2' ], [ 'F', '>=', 3, 10 ] ] ],
);

# update ... set v2 = -1 where (k1, k2) >= ('k1_3', '') and v2 = 10
exec_multi(
  4, "FILTER",
  [ 3, '>=', [ 'k1_1', '' ], 1000, 0, 'U', [ -1 ],
    [ [ 'F', '=', 1, 10 ] ] ],
);

exec_multi(
  4, "VAL",
  [ 1, '>=', [ '', '' ], 1000, 0 ],
);
# all

exit 0;

sub exec_single {
  my ($width, $mess, $req) = @_;
  print "$mess\n";
  my $res = $hs->execute_single(@$req);
  {
    my $code = shift(@$res);
    print "code=$code\n";
    my $i = 0;
    for my $fld (@$res) {
      print "[$fld]";
      if (++$i >= $width) {
	print "\n";
	$i = 0;
      }
    }
    print "\n";
  }
}

sub exec_multi {
  my $width = shift(@_);
  my $mess = shift(@_);
  print "$mess\n";
  my $mres = $hs->execute_multi(\@_);
  for my $res (@$mres) {
    my $code = shift(@$res);
    print "code=$code\n";
    my $i = 0;
    for my $fld (@$res) {
      print "[$fld]";
      if (++$i >= $width) {
	print "\n";
	$i = 0;
      }
    }
    print "\n";
  }
}

