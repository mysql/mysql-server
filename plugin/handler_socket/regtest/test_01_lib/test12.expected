VAL
code=0
[k1_0][k2_0][0][0]
[k1_0][k2_1][1][0]
[k1_0][k2_2][2][0]
[k1_0][k2_3][3][0]
[k1_0][k2_4][4][0]
[k1_0][k2_5][5][0]
[k1_0][k2_6][6][0]
[k1_0][k2_7][7][0]
[k1_0][k2_8][8][0]
[k1_0][k2_9][9][0]
[k1_1][k2_0][1][0]
[k1_1][k2_1][2][1]
[k1_1][k2_2][3][2]
[k1_1][k2_3][4][3]
[k1_1][k2_4][5][4]
[k1_1][k2_5][6][5]
[k1_1][k2_6][7][6]
[k1_1][k2_7][8][7]
[k1_1][k2_8][9][8]
[k1_1][k2_9][10][9]
[k1_2][k2_0][2][0]
[k1_2][k2_1][3][2]
[k1_2][k2_2][4][4]
[k1_2][k2_3][5][6]
[k1_2][k2_4][6][8]
[k1_2][k2_5][7][10]
[k1_2][k2_6][8][12]
[k1_2][k2_7][9][14]
[k1_2][k2_8][10][16]
[k1_2][k2_9][11][18]
[k1_3][k2_0][3][0]
[k1_3][k2_1][4][3]
[k1_3][k2_2][5][6]
[k1_3][k2_3][6][9]
[k1_3][k2_4][7][12]
[k1_3][k2_5][8][15]
[k1_3][k2_6][9][18]
[k1_3][k2_7][10][21]
[k1_3][k2_8][11][24]
[k1_3][k2_9][12][27]
[k1_4][k2_0][4][0]
[k1_4][k2_1][5][4]
[k1_4][k2_2][6][8]
[k1_4][k2_3][7][12]
[k1_4][k2_4][8][16]
[k1_4][k2_5][9][20]
[k1_4][k2_6][10][24]
[k1_4][k2_7][11][28]
[k1_4][k2_8][12][32]
[k1_4][k2_9][13][36]
[k1_5][k2_0][5][0]
[k1_5][k2_1][6][5]
[k1_5][k2_2][7][10]
[k1_5][k2_3][8][15]
[k1_5][k2_4][9][20]
[k1_5][k2_5][10][25]
[k1_5][k2_6][11][30]
[k1_5][k2_7][12][35]
[k1_5][k2_8][13][40]
[k1_5][k2_9][14][45]
[k1_6][k2_0][6][0]
[k1_6][k2_1][7][6]
[k1_6][k2_2][8][12]
[k1_6][k2_3][9][18]
[k1_6][k2_4][10][24]
[k1_6][k2_5][11][30]
[k1_6][k2_6][12][36]
[k1_6][k2_7][13][42]
[k1_6][k2_8][14][48]
[k1_6][k2_9][15][54]
[k1_7][k2_0][7][0]
[k1_7][k2_1][8][7]
[k1_7][k2_2][9][14]
[k1_7][k2_3][10][21]
[k1_7][k2_4][11][28]
[k1_7][k2_5][12][35]
[k1_7][k2_6][13][42]
[k1_7][k2_7][14][49]
[k1_7][k2_8][15][56]
[k1_7][k2_9][16][63]
[k1_8][k2_0][8][0]
[k1_8][k2_1][9][8]
[k1_8][k2_2][10][16]
[k1_8][k2_3][11][24]
[k1_8][k2_4][12][32]
[k1_8][k2_5][13][40]
[k1_8][k2_6][14][48]
[k1_8][k2_7][15][56]
[k1_8][k2_8][16][64]
[k1_8][k2_9][17][72]
[k1_9][k2_0][9][0]
[k1_9][k2_1][10][9]
[k1_9][k2_2][11][18]
[k1_9][k2_3][12][27]
[k1_9][k2_4][13][36]
[k1_9][k2_5][14][45]
[k1_9][k2_6][15][54]
[k1_9][k2_7][16][63]
[k1_9][k2_8][17][72]
[k1_9][k2_9][18][81]

FILTER
code=0
[k1_0][k2_5][5][0]
[k1_1][k2_5][6][5]
[k1_2][k2_5][7][10]
[k1_3][k2_5][8][15]
[k1_4][k2_5][9][20]
[k1_5][k2_5][10][25]
[k1_6][k2_5][11][30]
[k1_7][k2_5][12][35]
[k1_8][k2_5][13][40]
[k1_9][k2_5][14][45]

FILTER
code=0
[k1_0][k2_5][5][0]
[k1_1][k2_5][6][5]
[k1_2][k2_5][7][10]
[k1_3][k2_5][8][15]
[k1_4][k2_5][9][20]
[k1_5][k2_5][10][25]
[k1_6][k2_5][11][30]
[k1_7][k2_5][12][35]
[k1_8][k2_5][13][40]
[k1_9][k2_5][14][45]

FILTER
code=0
[k1_0][k2_3][3][0]
[k1_1][k2_2][3][2]
[k1_2][k2_1][3][2]
[k1_3][k2_0][3][0]

FILTER
code=0
[k1_1][k2_0][1][0]
[k1_1][k2_1][2][1]
[k1_1][k2_2][3][2]
[k1_1][k2_3][4][3]
[k1_1][k2_4][5][4]
[k1_1][k2_5][6][5]
[k1_1][k2_6][7][6]
[k1_1][k2_7][8][7]
[k1_1][k2_8][9][8]
[k1_1][k2_9][10][9]
[k1_2][k2_0][2][0]
[k1_2][k2_1][3][2]
[k1_2][k2_2][4][4]
[k1_2][k2_3][5][6]
[k1_2][k2_4][6][8]
[k1_2][k2_5][7][10]
[k1_2][k2_6][8][12]
[k1_2][k2_7][9][14]
[k1_2][k2_8][10][16]
[k1_2][k2_9][11][18]

FILTER
code=0
[k1_2][k2_5][7][10]
[k1_2][k2_6][8][12]
[k1_2][k2_7][9][14]
[k1_2][k2_8][10][16]
[k1_2][k2_9][11][18]

FILTER
code=0
[2]
VAL
code=0
[k1_0][k2_0][0][0]
[k1_0][k2_1][1][0]
[k1_0][k2_2][2][0]
[k1_0][k2_3][3][0]
[k1_0][k2_4][4][0]
[k1_0][k2_5][5][0]
[k1_0][k2_6][6][0]
[k1_0][k2_7][7][0]
[k1_0][k2_8][8][0]
[k1_0][k2_9][9][0]
[k1_1][k2_0][1][0]
[k1_1][k2_1][2][1]
[k1_1][k2_2][3][2]
[k1_1][k2_3][4][3]
[k1_1][k2_4][5][4]
[k1_1][k2_5][6][5]
[k1_1][k2_6][7][6]
[k1_1][k2_7][8][7]
[k1_1][k2_8][9][8]
[k1_1][k2_9][10][9]
[k1_2][k2_0][2][0]
[k1_2][k2_1][3][2]
[k1_2][k2_2][4][4]
[k1_2][k2_3][5][6]
[k1_2][k2_4][6][8]
[k1_2][k2_5][-1][10]
[k1_2][k2_6][8][12]
[k1_2][k2_7][9][14]
[k1_2][k2_8][10][16]
[k1_2][k2_9][11][18]
[k1_3][k2_0][3][0]
[k1_3][k2_1][4][3]
[k1_3][k2_2][5][6]
[k1_3][k2_3][6][9]
[k1_3][k2_4][7][12]
[k1_3][k2_5][8][15]
[k1_3][k2_6][9][18]
[k1_3][k2_7][10][21]
[k1_3][k2_8][11][24]
[k1_3][k2_9][12][27]
[k1_4][k2_0][4][0]
[k1_4][k2_1][5][4]
[k1_4][k2_2][6][8]
[k1_4][k2_3][7][12]
[k1_4][k2_4][8][16]
[k1_4][k2_5][9][20]
[k1_4][k2_6][10][24]
[k1_4][k2_7][11][28]
[k1_4][k2_8][12][32]
[k1_4][k2_9][13][36]
[k1_5][k2_0][5][0]
[k1_5][k2_1][6][5]
[k1_5][k2_2][-1][10]
[k1_5][k2_3][8][15]
[k1_5][k2_4][9][20]
[k1_5][k2_5][10][25]
[k1_5][k2_6][11][30]
[k1_5][k2_7][12][35]
[k1_5][k2_8][13][40]
[k1_5][k2_9][14][45]
[k1_6][k2_0][6][0]
[k1_6][k2_1][7][6]
[k1_6][k2_2][8][12]
[k1_6][k2_3][9][18]
[k1_6][k2_4][10][24]
[k1_6][k2_5][11][30]
[k1_6][k2_6][12][36]
[k1_6][k2_7][13][42]
[k1_6][k2_8][14][48]
[k1_6][k2_9][15][54]
[k1_7][k2_0][7][0]
[k1_7][k2_1][8][7]
[k1_7][k2_2][9][14]
[k1_7][k2_3][10][21]
[k1_7][k2_4][11][28]
[k1_7][k2_5][12][35]
[k1_7][k2_6][13][42]
[k1_7][k2_7][14][49]
[k1_7][k2_8][15][56]
[k1_7][k2_9][16][63]
[k1_8][k2_0][8][0]
[k1_8][k2_1][9][8]
[k1_8][k2_2][10][16]
[k1_8][k2_3][11][24]
[k1_8][k2_4][12][32]
[k1_8][k2_5][13][40]
[k1_8][k2_6][14][48]
[k1_8][k2_7][15][56]
[k1_8][k2_8][16][64]
[k1_8][k2_9][17][72]
[k1_9][k2_0][9][0]
[k1_9][k2_1][10][9]
[k1_9][k2_2][11][18]
[k1_9][k2_3][12][27]
[k1_9][k2_4][13][36]
[k1_9][k2_5][14][45]
[k1_9][k2_6][15][54]
[k1_9][k2_7][16][63]
[k1_9][k2_8][17][72]
[k1_9][k2_9][18][81]

